File: app/android/app/src/main/java/com/nativestream/android/data/remote/ApiClient.kt
// app/src/main/java/com/nativestream/android/data/remote/ApiClient.kt
//
// API Client (Ktor)
// Mirrors APIClient.swift actor exactly — same endpoints, same error mapping,
// same timeout values. Server URL is configurable (not hardcoded) because
// Android connects over LAN, not localhost.
//
// Inject via Hilt (see AppModule.kt). Do not instantiate directly.

package com.nativestream.android.data.remote
import io.ktor.client.engine.android.Android


import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.plugins.logging.LogLevel
import io.ktor.client.plugins.logging.Logger
import io.ktor.client.plugins.logging.Logging
import io.ktor.client.request.delete
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.client.request.put
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json
import android.util.Log
import javax.inject.Singleton

private const val TAG = "ApiClient"
private const val REQUEST_TIMEOUT_MS  = 10_000L
private const val RESOURCE_TIMEOUT_MS = 30_000L
private const val UNMATCHED_DEFAULT_LIMIT = 50

@Singleton
class ApiClient  constructor() {

    // ── Base URL — set during onboarding / settings change ───────────────────

    @Volatile private var baseUrl: String = "http://192.168.1.42:8888"

    fun setBaseUrl(url: String) {
        baseUrl = url.trimEnd('/')
    }

    // ── Ktor HTTP client ──────────────────────────────────────────────────────

    private val httpClient = HttpClient(Android) {
        install(ContentNegotiation) {
            json(Json {
                ignoreUnknownKeys = true
                isLenient         = true
            })
        }
        install(Logging) {
            logger = object : Logger {
                override fun log(message: String) { Log.d(TAG, message) }
            }
            level = LogLevel.INFO
        }
        engine {
            connectTimeout = REQUEST_TIMEOUT_MS.toInt()
            socketTimeout  = RESOURCE_TIMEOUT_MS.toInt()
        }
    }

    // ── Health ────────────────────────────────────────────────────────────────

    suspend fun health(): HealthResponse =
        get("api/health")

    // ── Playlist & EPG (raw bytes — consumed by parsers in AND-005 / AND-006) ─

    suspend fun playlistData(): ByteArray =
        rawGet("playlist.m3u")

    suspend fun epgData(): ByteArray =
        rawGet("epg.xml")

    // ── Channels ──────────────────────────────────────────────────────────────

    suspend fun listChannels(): List<ChannelResponse> {
        val response: ChannelListResponse = get("api/channels")
        return response.channels
    }

    suspend fun getChannel(id: String): ChannelDetailResponse =
        get("api/channels/$id")

    suspend fun createChannel(request: CreateChannelRequest): ChannelDetailResponse =
        post("api/channels", request)

    suspend fun updateChannel(id: String, request: UpdateChannelRequest) {
        put<StatusResponse>("api/channels/$id", request)
    }

    suspend fun deleteChannel(id: String) {
        executeDelete("api/channels/$id")
    }

    // ── Probe ─────────────────────────────────────────────────────────────────

    suspend fun triggerProbe() {
        post<StatusResponse>("api/probe", EmptyBody())
    }

    // ── Discovery ─────────────────────────────────────────────────────────────

    suspend fun discoveryStatus(): DiscoveryStatusResponse =
        get("api/discovery/status")

    suspend fun triggerDiscovery() {
        post<StatusResponse>("api/discovery/run", EmptyBody())
    }

    suspend fun unmatchedLinks(limit: Int = UNMATCHED_DEFAULT_LIMIT): UnmatchedResponse =
        get("api/discovery/unmatched?limit=$limit")

    suspend fun assignUnmatchedLink(channelId: String, streamUrl: String) {
        updateChannel(channelId, UpdateChannelRequest(streamUrl = streamUrl))
    }

    // ── HTTP primitives ───────────────────────────────────────────────────────

    private suspend inline fun <reified T> get(path: String): T =
        executeAndDecode { httpClient.get(resolve(path)) }

    private suspend fun rawGet(path: String): ByteArray =
        wrapNetworkErrors(path) {
            val response = httpClient.get(resolve(path))
            guardSuccess(response)
            response.body()
        }

    private suspend inline fun <reified T> post(path: String, body: Any): T =
        executeAndDecode {
            httpClient.post(resolve(path)) {
                contentType(ContentType.Application.Json)
                setBody(body)
            }
        }

    private suspend inline fun <reified T> put(path: String, body: Any): T =
        executeAndDecode {
            httpClient.put(resolve(path)) {
                contentType(ContentType.Application.Json)
                setBody(body)
            }
        }

    private suspend fun executeDelete(path: String) {
        wrapNetworkErrors(path) {
            val response = httpClient.delete(resolve(path))
            guardSuccess(response)
        }
    }

    private suspend inline fun <reified T> executeAndDecode(
        crossinline block: suspend () -> HttpResponse,
    ): T = wrapNetworkErrors("") {
        val response = block()
        guardSuccess(response)
        try {
            response.body<T>()
        } catch (cause: Exception) {
            throw ApiError.DecodingFailed(cause)
        }
    }

    private suspend fun <T> wrapNetworkErrors(path: String, block: suspend () -> T): T =
        try {
            block()
        } catch (apiError: ApiError) {
            throw apiError
        } catch (cause: Exception) {
            throw ApiError.ServerUnreachable(resolve(path))
        }

    private suspend fun guardSuccess(response: HttpResponse) {
        if (!response.status.isSuccess()) {
            val body = runCatching { response.bodyAsText() }.getOrNull()
            throw ApiError.HttpError(response.status.value, body)
        }
    }

    private fun resolve(path: String): String = "$baseUrl/$path"
}

File: apps/macos/NativeStream/NativeStream/API/APIClient.swift
// APIClient.swift
// Typed networking layer for all NativeStream Server endpoints.
// All methods are async throws. Errors are mapped to APIError.

import Foundation
import SwiftProtobuf
import SdkGenSwift


// MARK: - APIClient

actor APIClient {

    static let shared = APIClient()

    private var baseURL: URL
    private let session: URLSession

    private static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .iso8601
        return d
    }()

    private static let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.dateEncodingStrategy = .iso8601
        return e
    }()
    
    private func rawProtoBody(
        method: String, path: String, message: any SwiftProtobuf.Message
    ) async throws -> Data {
        let url = try resolve(path)
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        var options = JSONEncodingOptions()
        options.preserveProtoFieldNames = true
        req.httpBody = try message.jsonUTF8Data(options: options)
        return try await execute(req)
    }

    init(
        baseURL: URL = URL(string: "http://localhost:8888")!,
        protocolClasses: [AnyClass]? = nil
    ) {
        self.baseURL = baseURL
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest  = 10
        config.timeoutIntervalForResource = 30
        config.urlCache = URLCache(
            memoryCapacity: 10 * 1024 * 1024,  // 10 MB
            diskCapacity:   50 * 1024 * 1024,  // 50 MB
            diskPath:       "nativestream_api_cache"
        )
        config.requestCachePolicy = .useProtocolCachePolicy
        if let protocolClasses {
            config.protocolClasses = protocolClasses
        }
        self.session = URLSession(configuration: config)
    }

    func setBaseURL(_ url: URL) {
        baseURL = url
    }

    // MARK: - Health

    func health() async throws -> Stream_V1_HealthResponse {
        let data = try await rawGet("api/health")
        do {
            return try Stream_V1_HealthResponse(jsonUTF8Data: data)
        } catch {
            throw APIError.decodingFailed(error)
        }
    }

    // MARK: - Playlist & EPG (raw Data — parsed by existing parsers)

    func playlistData() async throws -> Data {
        try await rawGet("playlist.m3u")
    }
    
    func probePlaylistForEpg(url: URL) async -> URL? {
        guard let data = try? await fetchRawURL(url),
              let text = String(data: data.prefix(8192), encoding: .utf8) ??
                         String(data: data.prefix(8192), encoding: .isoLatin1) else { return nil }

        let header = text.components(separatedBy: .newlines).first ?? text

        for key in ["url-tvg", "x-tvg-url"] {
            if let value = M3UParser.extractAttribute(key, from: header) {
                return URL(string: value)
            }
        }
        return nil
    }
    
    func fetchRawURL(_ url: URL) async throws -> Data {
        var req = URLRequest(url: url)
        req.cachePolicy = .useProtocolCachePolicy
        return try await execute(req)
    }

    func epgData() async throws -> Data {
        try await rawGet("epg.xml")
    }

    // MARK: - Channels

    func listChannels() async throws -> [Stream_V1_ChannelResponse] {
        let data = try await rawGet("api/channels")
        do {
            return try Stream_V1_ChannelListResponse(jsonUTF8Data: data).channels
        } catch {
            throw APIError.decodingFailed(error)
        }
    }

    func getChannel(id: String) async throws -> Stream_V1_ChannelDetailResponse {
        let data = try await rawGet("api/channels/\(id)")
        do {
            return try Stream_V1_ChannelDetailResponse(jsonUTF8Data: data)
        } catch {
            throw APIError.decodingFailed(error)
        }
    }

    func createChannel(_ req: Stream_V1_CreateChannelRequest) async throws -> Stream_V1_ChannelDetailResponse {
        let data = try await rawProtoBody(method: "POST", path: "api/channels", message: req)
        do {
            return try Stream_V1_ChannelDetailResponse(jsonUTF8Data: data)
        } catch {
            throw APIError.decodingFailed(error)
        }
    }

    func updateChannel(id: String, _ req: Stream_V1_UpdateChannelRequest) async throws {
        let data = try await rawProtoBody(method: "PUT", path: "api/channels/\(id)", message: req)
        do {
            _ = try Stream_V1_StatusResponse(jsonUTF8Data: data)
        } catch {
            throw APIError.decodingFailed(error)
        }
    }

    func deleteChannel(id: String) async throws {
        let url = try resolve("api/channels/\(id)")
        var req = URLRequest(url: url)
        req.httpMethod = "DELETE"
        let data = try await execute(req)
        do {
            _ = try Stream_V1_StatusResponse(jsonUTF8Data: data)
        } catch {
            throw APIError.decodingFailed(error)
        }
    }
    
    // MARK: - Discovery

    func discoveryStatus() async throws -> Stream_V1_DiscoveryStatusResponse {
        let data = try await rawGet("api/discovery/status")
        do {
            return try Stream_V1_DiscoveryStatusResponse(jsonUTF8Data: data)
        } catch {
            throw APIError.decodingFailed(error)
        }
    }

    func triggerDiscovery() async throws {
        let url = try resolve("api/discovery/run")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        let data = try await execute(req)
        do {
            _ = try Stream_V1_StatusResponse(jsonUTF8Data: data)
        } catch {
            throw APIError.decodingFailed(error)
        }
    }

    func unmatchedLinks(limit: Int = 50) async throws -> Stream_V1_UnmatchedResponse {
        let data = try await rawGet("api/discovery/unmatched?limit=\(limit)")
        do {
            return try Stream_V1_UnmatchedResponse(jsonUTF8Data: data)
        } catch {
            throw APIError.decodingFailed(error)
        }
    }

    func assignUnmatchedLink(channelID: String, url: String) async throws {
        var req = Stream_V1_UpdateChannelRequest()
        req.streamURL = url
        try await updateChannel(id: channelID, req)
    }

    // MARK: - Probe

    func triggerProbe() async throws {
        let _: StatusResponse = try await post("api/probe", body: EmptyBody())
    }
    
    // MARK: - Proxy config

    func getProxyEnabled() async throws -> Bool {
        let data = try await rawGet("api/proxy/config")
        do {
            return try Stream_V1_ProxyConfigResponse(jsonUTF8Data: data).enabled
        } catch {
            throw APIError.decodingFailed(error)
        }
    }
    
    func setProxyEnabled(_ enabled: Bool) async throws {
        var req = Stream_V1_UpdateProxyConfigRequest()
        req.enabled = enabled
        let data = try await rawProtoBody(method: "PUT", path: "api/proxy/config", message: req)
        do {
            _ = try Stream_V1_ProxyConfigResponse(jsonUTF8Data: data)
        } catch {
            throw APIError.decodingFailed(error)
        }
    }

    // MARK: - Private HTTP primitives

    private func get<T: Decodable>(_ path: String) async throws -> T {
        let data = try await rawGet(path)
        return try decode(T.self, from: data)
    }

    private func post<Body: Encodable, Response: Decodable>(
        _ path: String, body: Body
    ) async throws -> Response {
        let data = try await rawBody(method: "POST", path: path, body: body)
        return try decode(Response.self, from: data)
    }

    private func put<Body: Encodable, Response: Decodable>(
        _ path: String, body: Body
    ) async throws -> Response {
        let data = try await rawBody(method: "PUT", path: path, body: body)
        return try decode(Response.self, from: data)
    }

    private func delete<Response: Decodable>(_ path: String) async throws -> Response {
        let url = try resolve(path)
        var req = URLRequest(url: url)
        req.httpMethod = "DELETE"
        let data = try await execute(req)
        return try decode(Response.self, from: data)
    }

    private func rawGet(_ path: String) async throws -> Data {
        let url = try resolve(path)
        var req = URLRequest(url: url)
        req.cachePolicy = .useProtocolCachePolicy
        req.setValue("max-age=7200, public", forHTTPHeaderField: "Cache-Control")
        return try await execute(req)
    }

    private func rawBody<Body: Encodable>(
        method: String, path: String, body: Body
    ) async throws -> Data {
        let url = try resolve(path)
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try APIClient.encoder.encode(body)
        return try await execute(req)
    }

    private func execute(_ request: URLRequest) async throws -> Data {
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw APIError.serverUnreachable(request.url ?? baseURL)
        }
        guard let http = response as? HTTPURLResponse else {
            throw APIError.serverUnreachable(request.url ?? baseURL)
        }
        guard (200...299).contains(http.statusCode) else {
            let msg = String(data: data, encoding: .utf8)
            throw APIError.httpError(http.statusCode, msg)
        }
        return data
    }

    private func resolve(_ path: String) throws -> URL {
        guard let url = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
            throw APIError.invalidURL(path)
        }
        return url
    }

    private func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T {
        do {
            return try APIClient.decoder.decode(type, from: data)
        } catch {
            throw APIError.decodingFailed(error)
        }
    }
}

// MARK: - Helpers

private struct EmptyBody: Encodable {}
